### PR TITLE
Fix various concurrency issues in transport

### PR DESCRIPTION
**Pull request by danielmitterdorfer**
Due to various issues (most notably a missing happens-before edge 
between socket accept and channel close in MockTcpTransport), 
MockTcpTransportTests sometimes did not terminate.

With this commit we fix various concurrency issues that led to
this hanging test.

Failing example build: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-os-compatibility/os=oraclelinux/835/console

### CODE REVIEW

**Comment by danielmitterdorfer** [comment]
@bleskes can you please review this?

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java [hunk 1]
```diff
@@ -470,11 +469,13 @@ protected final void disconnectFromNodeChannel(final Channel channel, final Exce
     }
 
     protected Channel nodeChannel(DiscoveryNode node, TransportRequestOptions options) throws ConnectTransportException {
-        NodeChannels nodeChannels = connectedNodes.get(node);
-        if (nodeChannels == null) {
-            throw new NodeNotConnectedException(node, "Node not connected");
+        try (Releasable ignored = connectionLock.acquire(node.getId())) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: why is this needed? connectedNodes is a concurrent map? note that this is called with every request

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java [hunk 2]
```diff
@@ -768,19 +769,19 @@ protected final void doStop() {
...
-
                 for (Map.Entry<String, List<Channel>> entry : serverChannels.entrySet()) {
                     try {
                         closeChannels(entry.getValue());
                     } catch (Exception e) {
                         logger.debug("Error closing serverChannel for profile [{}]", e, entry.getKey());
                     }
                 }
+
+                for (Iterator<NodeChannels> it = connectedNodes.values().iterator(); it.hasNext(); ) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: why did you move this? can you leave a comment if it's important? also, we do it again in the finally close of the next statement. Did you find out why it needs to be done twice? if so, please add a comment :)

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java [hunk 3]
```diff
@@ -919,6 +919,7 @@ public void sendRequest(final DiscoveryNode node, final long requestId, final St
                     transportServiceAdapter.onRequestSent(node, requestId, action, request, finalOptions);
                 }
             };
+            Channel targetChannel = nodeChannel(node, options);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: why did you move this? again, original idea is to have the serialization saved when we are disconnected from a node, which most code doesn't check for in advance.

File: test/framework/src/main/java/org/elasticsearch/transport/MockTcpTransport.java [hunk 4]
```diff
@@ -261,20 +261,50 @@ public long serverOpen() {
         private final CancellableThreads cancellableThreads = new CancellableThreads();
         private final Closeable onClose;
 
+        /**
+         * Constructs a new MockChannel instance intended as client.
```
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/transport/MockTcpTransport.java]: not sure exactly what you mean with "client". This is used both when this node connects to another node and when this nodes receives a connection from another one.

File: test/framework/src/main/java/org/elasticsearch/transport/MockTcpTransport.java [hunk 5]
```diff
@@ -261,20 +261,50 @@ public long serverOpen() {
...
+            this.profile = profile;
+            this.activeChannel = null;
+            this.onClose = null;
+        }
+
         public void accept(Executor executor) throws IOException {
             while (isOpen.get()) {
                 Socket accept = serverSocket.accept();
                 configureSocket(accept);
                 MockChannel mockChannel = new MockChannel(accept, localAddress, profile, workerChannels::remove);
```
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/transport/MockTcpTransport.java]: we may still leak things on exceptions. How about:

```
            while (isOpen.get()) {
                Socket incomingSocket = serverSocket.accept();
                MockChannel incomingChannel = null;
                try {
                    configureSocket(incomingSocket);
                    incomingChannel = new MockChannel(incomingSocket, localAddress, profile, workerChannels::remove);
                    //establish a happens-before edge between closing and accepting a new connection
                    synchronized (this) {
                        if (isOpen.get()) {
                            workerChannels.put(incomingChannel, Boolean.TRUE);
                            // the channel is properly registered and will be cleared by the close code.
                            incomingChannel = null;
                            incomingSocket = null;
                            // this spawns a new thread immediately, so OK under lock
                            incomingChannel.loopRead(executor);
                        }
                    }
                } finally {
                    IOUtils.closeWhileHandlingException(incomingSocket, incomingChannel);
                }
            }
```

File: test/framework/src/main/java/org/elasticsearch/transport/MockTcpTransport.java [hunk 6]
```diff
@@ -294,26 +324,21 @@ public void onFailure(Exception e) {
                 @Override
                 protected void doRun() throws Exception {
                     StreamInput input = new InputStreamStreamInput(new BufferedInputStream(activeChannel.getInputStream()));
-                    while (isOpen.get()) {
+                    while (isOpen.get() && !Thread.currentThread().isInterrupted()) {
```
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/transport/MockTcpTransport.java]: who interrupts this thread that didn't call the close method and thus cancellableThreads.cancel()?

**Comment by bleskes** [comment]
left some comments. Good work. Did you manage to reproduce the failure with sleeps and see that fix fixes it ?  I think it's good to also link to an example build failure in the PR/commit description.

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java (hunk 1 above)
Comment (danielmitterdorfer) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: This is needed to establish a happens-before edge between `#nodeChannel()` and `#disconnectFromNode()`. Without the lock a connect can sneak in in the middle of a disconnect.

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: connect and disconnect are arranged in their proper methods: see TcpTransport#connectToNode(org.elasticsearch.cluster.node.DiscoveryNode, boolean)

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java (hunk 2 above)
Comment (danielmitterdorfer) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: > why did you move this?

The first thing I wanted to do is stopping this transport from accepting any incoming connections and only when we're done with that I wanted to close connected nodes. This change is not really necessary but to me this is the cleaner option. I can add a comment.

> Did you find out why it needs to be done twice? if so, please add a comment :)

It's not needed: We're operating under a lock (which is also used by `#connectToNode()`(!)) and we've just capped all incoming and outgoing connections. It was just a leftover and I'll remove it in my next commit.

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java (hunk 3 above)
Comment (danielmitterdorfer) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: I wanted to reduce the scope (or rather the lifetime) of this variable and bring it as close as possible to its usage. It's not needed anywhere before this point and t.b.h. we can even inline it and get rid of the variable.

File: test/framework/src/main/java/org/elasticsearch/transport/MockTcpTransport.java (hunk 4 above)
Comment (danielmitterdorfer) [inline test/framework/src/main/java/org/elasticsearch/transport/MockTcpTransport.java]: You're right. I've changed the Javadoc comments for both constructors now.

File: test/framework/src/main/java/org/elasticsearch/transport/MockTcpTransport.java (hunk 5 above)
Comment (danielmitterdorfer) [inline test/framework/src/main/java/org/elasticsearch/transport/MockTcpTransport.java]: Good catch! I'll change that (though we should not close socket and channel unconditionally (because if we're not closed they must stay open).

Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/transport/MockTcpTransport.java]: it seems an extra comment in the finally is needed :)

File: test/framework/src/main/java/org/elasticsearch/transport/MockTcpTransport.java (hunk 6 above)
Comment (danielmitterdorfer) [inline test/framework/src/main/java/org/elasticsearch/transport/MockTcpTransport.java]: Our test infrastructure for example. Admittedly, this is a *really* small window and very unlikely, but it is possible nevertheless.

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java (hunk 3 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: as I said, I think doing it early to catch node close quickly it's intentional (though not a big deal). Can we keep it as is then?

File: test/framework/src/main/java/org/elasticsearch/transport/MockTcpTransport.java (hunk 6 above)
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/transport/MockTcpTransport.java]: but if that happens and we make cancellable thread not clear the interrupt ( :)  ) it won't be needed, right? if so ++ to our other plan.

Can you please leave a comment?

**Comment by danielmitterdorfer** [comment]
> Did you manage to reproduce the failure with sleeps and see that fix fixes it ? 

I added a sleep right after `Socket incomingSocket = serverSocket.accept();` in `MockTcpTransport` to provoke the issue. With this change I could reproduce the issue in *almost* every run.

With the changes the code ran smoothly, i.e. without any pauses at all for several 100.000 iterations in a row (in addition to the complete stall, the code hang previously also for a around 10 seconds (which has been caused by the timeout in `MockTcpTransport#stopInternal())` on pool shutdown).

> I think it's good to also link to an example build failure in the PR/commit description. 

Thanks for the hint. I'll add that soon.

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java (hunk 1 above)
Comment (danielmitterdorfer) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: That's correct. But I was worried between the HB-edge between a new request incoming (-> `#nodeChannel()') and a concurrent disconnect (`#disconnectFromNode`).

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java (hunk 3 above)
Comment (danielmitterdorfer) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: Sure. I'll move it back up.

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java (hunk 1 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: but that can happen anyway, as this method returns a channel and then it's up for people to deal with it. double checking - you saw that this map is a concurrent one?

File: test/framework/src/main/java/org/elasticsearch/transport/MockTcpTransport.java (hunk 5 above)
Comment (danielmitterdorfer) [inline test/framework/src/main/java/org/elasticsearch/transport/MockTcpTransport.java]: Definitely. Added!

File: test/framework/src/main/java/org/elasticsearch/transport/MockTcpTransport.java (hunk 6 above)
Comment (danielmitterdorfer) [inline test/framework/src/main/java/org/elasticsearch/transport/MockTcpTransport.java]: You're right. I forgot about the custom interruption policy in `CancellableThreads`. I've added a comment.

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java (hunk 1 above)
Comment (danielmitterdorfer) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: Yes, I am fully aware that this is a concurrent map and this was never my issue. Your last comment has convinced me to restore the original code though. Thanks for your persistence and the explanation. :)

**Comment by danielmitterdorfer** [comment]
I've pushed another commit that addresses all your comments.

File: test/framework/src/main/java/org/elasticsearch/transport/MockTcpTransport.java [hunk 7]
```diff
@@ -294,26 +333,23 @@ public void onFailure(Exception e) {
                 @Override
                 protected void doRun() throws Exception {
                     StreamInput input = new InputStreamStreamInput(new BufferedInputStream(activeChannel.getInputStream()));
+                    // as per interruption policy of CancellableThreads we don't need to check for interrupts.
```
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/transport/MockTcpTransport.java]: I think there is still a extremely small chance someone interrupts the thread before cancellable threads is called and since we clean up any interrupt flag over before we start, it won't block the read. So I think what you had is strictly speaking needed. However, for the testing infra part, I don't think we're required to honor that interrupt and we'll be killed anyway. I'm OK with the current code, but the comment is not correct.

Comment (danielmitterdorfer) [inline test/framework/src/main/java/org/elasticsearch/transport/MockTcpTransport.java]: Ok, then I revert the change because I think it's best we just exit the loop in case we're interrupted.

**Comment by bleskes** [comment]
LGTM. I thought I did long time ago.

**Comment by danielmitterdorfer** [comment]
@bleskes Thanks for your review. :)
